Lecture 1 solutions

bykbaiev opened the pull request:
### Solutions for Lecture 1

cc @chshersh @vrom911

Commit:
Complete homework exercises for Lecture 1

lecture-1-solutions

Commit:
Add type signatures for helper functions (lowerAndGreater)

lecture-1-solutions

Commit:
Simplify substring function

Omit string parameter (use partial application)

lecture-1-solutions

Commit:
Replace power operator (^) with multiplication (*) (sumOfSquares)

To not deal with Integers that power operator uses as it's parameters

lecture-1-solutions

Commit:
Remove extra brackets and use better names for parameters

Small refactoring to solve some lint warnings

lecture-1-solutions

Commit:
Simplify strSum function

Use composition of function instead of additional value

lecture-1-solutions

Commit:
Replace if conditions with guards (lowerAndGreater)

lecture-1-solutions

chshersh commented on src/Lecture1.hs:
```diff
@@ -63,7 +65,8 @@ sumOfSquares x y = error "TODO!"
 -- DON'T FORGET TO SPECIFY THE TYPE IN HERE
-lastDigit n = error "lastDigit: Not implemented!"
+lastDigit :: Int -> Int
+lastDigit n = abs n `mod` 10
```
Perfect 💯

chshersh commented on src/Lecture1.hs:
```diff
@@ -94,7 +100,8 @@ start position can be considered as zero (e.g. substring from the
 -}
-subString start end str = error "TODO"
+subString :: Int -> Int -> String -> String
+subString start end = drop start . take (end + 1)
```
This is a neat implementation! 😎 

I'm already happy with it 🤗 However, it's possible to implement it more efficiently by dropping first and taking the required number of elements after that. I think it will be more efficient. But this is because I know how `drop` and `take` are implemented internally 😅 These both functions traverse the list until they reach the point of stopping.

So, for example, if you're given a string of 100 elements and you want to have a substring from index 70 to 99, your existing implementation will do 170 operations (99 + 1 for taking the entire list and then 70 for dropping the prefix) while dropping 70 first and taking the remaining 30 will require only 100 operations total.

But of course, I haven't explained the internals of `take` and `drop` in the first lecture. And you're not expected to know such internal implementation details at this point. Therefore, your implementation is pretty elegant 🆒

bykbaiev replied:
Thanks a lot! It's definitely easier (I was quite lazy to skip that part because didn't want to calculate the number of characters I need to take with all that boundaries about negative indices :D)

chshersh commented on src/Lecture1.hs:
```diff
@@ -119,4 +127,15 @@ and lower than 6 elements (4, 5, 6, 7, 8 and 9).
 -}
-lowerAndGreater n list = error "TODO"
+lowerAndGreater :: Int -> [Int] -> String
+lowerAndGreater n list = show n ++ " is greater than " ++ (show . getLower 0) list ++ " elements and lower than " ++ (show . getGreater 0) list ++ " elements"
```
That's correct! 💯 

However, with this solution, you will walk through the list twice.
You can write it more performant with only one traversal of the list 🙂
It's desirable to write functions that traverse lists single time and consume elements only once. In that case, the composition of such functions will have nice streaming and lazy properties.

bykbaiev replied:
thanks! And do you use transducers in Haskell at all? The case is I work with ClojureScript so love them a lot 😄 And the fact that most of functions for working with collections are transducers is a nice bonus)

chshersh replied:
I'm not familiar with transducers in ClojureScript 🤔 
Due to lazy evaluation, every function that consumes lists in Haskell only once is automatically streaming and consumes constant memory if written properly 🙂

bykbaiev replied:
Thanks a lot for your comments! I reimplemented the solution. Will resolve this for now to not waste your time on re-check

chshersh reviewed (approved):
Congrats! The first homework is complete 👏🏼 
Hope you enjoyed the course so far. Keep up the great work, and feel free to ask any questions if you have 👌🏼

Commit:
Refactor subStr function

To decrease number of computations at first drop first extra chars and then take what needed

lecture-1-solutions

Commit:
Refactor lowerAndGreater function

Replace 2 helpers with only one to decrease number of computation (to traverse the list single time)

lecture-1-solutions
